base/exporter.go

```diff
@@ -0,0 +1,714 @@
+	SyncFunctionCount   *SgwIntStat `json:"sync_function_count"`
+	SyncFunctionTime    *SgwIntStat `json:"sync_function_time"`
+	WriteProcessingTime *SgwIntStat `json:"write_processing_time"`
+}
+
+type DatabaseStats struct {
+	AbandonedSeqs           *SgwIntStat `json:"abandoned_seqs"`
+	ConflictWriteCount      *SgwIntStat `json:"conflict_write_count"`
+	Crc32MatchCount         *SgwIntStat `json:"crc32c_match_count"`
+	DCPCachingCount         *SgwIntStat `json:"dcp_caching_count"`
+	DCPCachingTime          *SgwIntStat `json:"dcp_caching_time"`
+	DCPReceivedCount        *SgwIntStat `json:"dcp_received_count"`
+	DCPReceivedTime         *SgwIntStat `json:"dcp_received_time"`
+	DocReadsBytesBlip       *SgwIntStat `json:"doc_reads_bytes_blip"`
+	DocWritesBytes          *SgwIntStat `json:"doc_writes_bytes"`
+	DocWritesBytesBlip      *SgwIntStat `json:"doc_writes_bytes_blip"`
+	DocWritesXattrBytes     *SgwIntStat `json:"doc_writes_xattr_bytes"`
+	HighSeqFeed             *SgwIntStat `json:"high_seq_feed"`
+	NumDocReadsBlip         *SgwIntStat `json:"num_doc_reads_blip"`
+	NumDocReadsRest         *SgwIntStat `json:"num_doc_reads_rest"`
+	NumDocWrites            *SgwIntStat `json:"num_doc_writes"`
+	NumReplicationsActive   *SgwIntStat `json:"num_replications_active"`
+	NumReplicationsTotal    *SgwIntStat `json:"num_replications_total"`
+	NumTombstonesCompacted  *SgwIntStat `json:"num_tombstones_compacted"`
+	SequenceAssignedCount   *SgwIntStat `json:"sequence_assigned_count"`
+	SequenceGetCount        *SgwIntStat `json:"sequence_get_count"`
+	SequenceIncrCount       *SgwIntStat `json:"sequence_incr_count"`
+	SequenceReleasedCount   *SgwIntStat `json:"sequence_released_count"`
+	SequenceReservedCount   *SgwIntStat `json:"sequence_reserved_count"`
+	WarnChannelsPerDocCount *SgwIntStat `json:"warn_channels_per_doc_count"`
+	WarnGrantsPerDocCount   *SgwIntStat `json:"warn_grants_per_doc_count"`
+	WarnXattrSizeCount      *SgwIntStat `json:"warn_xattr_size_count"`
+
+	// These can be cleaned up in future versions of SGW, implemented as maps to reduce amount of potential risk
+	// prior to Hydrogen release.
+	CacheFeedMapStats  *ExpVarMapWrapper `json:"cache_feed"`
+	ImportFeedMapStats *ExpVarMapWrapper `json:"import_feed"`
+}
+
+type ExpVarMapWrapper struct {
+	*expvar.Map
+}
+
+func (wrapper *ExpVarMapWrapper) MarshalJSON() ([]byte, error) {
+	return []byte(wrapper.String()), nil
+}
+
+type ReplicatorStats struct {
+	toplevelMap *expvar.Map
+}
+
+func (rs *ReplicatorStats) MarshalJSON() ([]byte, error) {
+	return []byte(rs.toplevelMap.String()), nil
+}
+
+func (rs *ReplicatorStats) Describe(ch chan<- *prometheus.Desc) {
+	return
+}
+
+func (rs *ReplicatorStats) Collect(ch chan<- prometheus.Metric) {
+	rs.toplevelMap.Do(func(value expvar.KeyValue) {
+		ch <- prometheus.MustNewConstMetric(
+			prometheus.NewDesc(prometheus.BuildFQName("sgw", "replication", "sgr_docs_checked_sent"), "sgr_docs_checked_sent", []string{"replication"}, nil),
```

Comment:
Why are these generated at collection time, unlike the rest of the stats?  I realize they aren't getting published to prometheus at a per-stat level, but is there a way to only generate the prometheus desc once?

Reply:
The work here is done on collect time as this is still based on an expvar map. This means we need to loop over it all at collection time as we don't know what stats are available prior to this. I have however moved the description generation to when the ReplicatorStats are initialized so that they are only generated once